src/runtime/manifest-primitive-parser/manifest-primitive-parser.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * @license
+ * Copyright (c) 2017 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Loader} from '../../platform/loader.js';
+import {parse as parser} from '../../gen/runtime/manifest-parser.js';
+
+import * as AstNode from '../manifest-ast-types/manifest-ast-nodes.js';
+import {Dictionary} from '../../utils/lib-utils.js';
+
+export type Ast = AstNode.All[];
+export {AstNode};
+
+interface ManifestLoadOptions {
+  registry?: Dictionary<Promise<Ast>>;
+  errors?: ManifestError[];
+}
+
+export interface ManifestParseOptions extends ManifestLoadOptions {
+  filename?: string;
+  loader?: Loader;
+}
+
+export enum ErrorSeverity {
+  Error = 'error',
+  Warning = 'warning'
+}
+
+export class ManifestError extends Error {
+  key: string;
+  location: AstNode.SourceLocation;
+  severity = ErrorSeverity.Error;
+  constructor(location: AstNode.SourceLocation, message: string) {
+    super(message);
+    this.location = location;
+  }
+}
+
+export class ManifestWarning extends ManifestError {
+  severity = ErrorSeverity.Warning;
+}
+
+export class ManifestPrimitiveParser {
+  static async load(path: string, loader: Loader, options: ManifestLoadOptions = {}): Promise<Ast> {
+    if (!loader) {
+      throw new Error('loader is required to load manifest for parsing');
+    }
+    // TODO(sjmiles):
+    // (1) IMO caching should be done at the Loader level
```

Comment:
Yes. I should have updated it to say that there is more than one module using a little custom caching dictionary, and it could be useful to factor out.